[DDP] Generalize activation checkpoint tests

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74252
* __->__ #74130

- Enables activation checkpointing tests to run for GLOO backend in addition to NCCL.
- Follow up changes will add support for non reentrant-based checkpointing and CheckpointWrapper.
- Changes AbstractDDPTest to CommonDDPTest to better indicate that it now has tests that are common to both backends
- 
Differential Revision: [D34281684](https://our.internmc.facebook.com/intern/diff/D34281684/)

rohan-varma committed:
[DDP] Generalize activation checkpoint tests

enable these tests to run for all dist backends not just nccl.

Differential Revision: [D34281684](https://our.internmc.facebook.com/intern/diff/D34281684/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Generalize activation checkpoint tests"


- Enables activation checkpointing tests to run for GLOO backend in addition to NCCL.
- Follow up changes will add support for non reentrant-based checkpointing and CheckpointWrapper.
- 
Differential Revision: [D34281684](https://our.internmc.facebook.com/intern/diff/D34281684/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Generalize activation checkpoint tests"


- Enables activation checkpointing tests to run for GLOO backend in addition to NCCL.
- Follow up changes will add support for non reentrant-based checkpointing and CheckpointWrapper.
- 
Differential Revision: [D34281684](https://our.internmc.facebook.com/intern/diff/D34281684/)

[ghstack-poisoned]

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+        return dist.FileStore(self.file_name, self.world_size)
+
+    def _get_process_group(self):
+        raise ValueError("To be implemented by child class")
```
nit: I think traditionally we raise `NotImplementedError()` if we are not using abstract base classes.

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+        allow_none_grads=False,
+    ):
+        # to reproduce the same training results
+        torch.cuda.set_device(self.rank)
```
Do the effects of `set_device()` persist after the test and possibly affect other tests?
If so, should we consider using `with torch.cuda.device(self.rank):` (context manager)?

rohan-varma replied:
It won't affect other tests because all tests are launched and run in their own subprocess, but in general it might be better to use the context manager.

Historically we've always used torch.cuda.set_device(self.rank) and it should be fine when the arg is self.rank because we usually do the GPU mapping that way.

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+                if not allow_none_grads:
+                    self.assertTrue(i.grad is not None)
+                    self.assertTrue(j.grad is not None)
+                self.assertEqual(i.grad, j.grad, rtol=1.3e-06, atol=5e-5)
```
General question: Do we prefer `self.assertEqual()` even for `torch.Tensor`s? I have been using the new `torch.testing.assert_close()` (see [here](https://github.com/pytorch/pytorch/issues/61844)).

rohan-varma replied:
I'm unsure when to use one over the other, although in this case I am directly porting the tests so don't want to change them too much.

pmeier replied:
Just saw this and thought I clarify a little: Internally, `TestCase.assertEqual` uses the same backend as `assert_close`, but is a lot more lenient. The original plan was to remove `TestCase.assertEqual` completely in favor of `assert_close`, but we dropped it, because a lot of our tests rely on the quirks.

If you write new tests, I would encourage you to use `assert_close`, since it is more strict by default and that eases the reviewing workflow.

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+        """
+        Similar to CheckpointTwiceModule but the weights are shared.
+        """
+        def __init__(self, use_reentrant=True):
```
This looks identical to `CheckpointTwiceModule`. Could you point out the weight sharing?

awgu replied:
Maybe it is missing the `self.l1.weight = self.l2.weight` line similar to L440?

rohan-varma replied:
Good catch! I'll add this in a follow up change - currently I am porting over the tests so don't want to change them in this diff iitself.

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+        DDP works as expected when layer is checkpointed only once.
+        """
+        process_group = self._get_process_group()
+        for use_bucket_view, static_graph in product((False, True), (False, True)):
```
Curious why `use_bucket_view` and `static_graph` are not parameterized. Is this to reduce setup/teardown costs?

awgu commented on test/distributed/test_c10d_common.py:
```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
+            )
+
+    def test_invalid_powerSGD_state(self):
+        for start_powerSGD_iter, use_error_feedback, warm_start in product(
```
Similar question here about not parameterizing.

awgu reviewed (commented):
Left some minor questions.

rohan-varma committed:
Update on "[DDP] Generalize activation checkpoint tests"


- Enables activation checkpointing tests to run for GLOO backend in addition to NCCL.
- Follow up changes will add support for non reentrant-based checkpointing and CheckpointWrapper.
- Changes AbstractDDPTest to CommonDDPTest to better indicate that it now has tests that are common to both backends
- 
Differential Revision: [D34281684](https://our.internmc.facebook.com/intern/diff/D34281684/)

[ghstack-poisoned]

awgu reviewed (approved):
LGTM!
